Add Worker/Queue to process expired DNS records

SerpentBytes opened the pull request:
## Description
Resolves #112 by adding Worker/Queue to do the following:
1. Query the database for expired DNS records
2. If Step 1. yields any results, delete DNS Records from Route53 and delete the record from DB (via `deleteDnsRequest`)
3. Send Notification 


Step 2 is not working as intended. Invoking `deleteDnsRequest` in the front-end/back-end fails with the following warning: 
```bash
warn: Invalid request: Expected exactly one of [AliasTarget, all of [TTL, and ResourceRecords], or TrafficPolicyInstanceId], but found none in Change with [Action=DELETE, Name=osd700-a1.user1.starchart.com, Type=A, SetIdentifier=null]
```
Possibly linked to the missing [TTL](https://github.com/boto/boto3/issues/3161) value.

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,105 @@
+        try {
+          logger.info(`process DNS record expiration`);
+          let dnsRecords = await getExpiredRecords(type);
+          if (dnsRecords?.length) {
```
This is unnecessary, you can `await Promise.all([])` and it will be ok.

humphd replied:
I'm confused why this is still here?

SerpentBytes replied:
> I'm confused why this is still here?

I don't know what you are referring to here.

humphd replied:
```ts
if (dnsRecords?.length) {
```

SerpentBytes replied:
Only try deleting if we get results back from a query, or is that just unnecessary since the job will fail?

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,105 @@
+          if (dnsRecords?.length) {
+            await Promise.all(
+              dnsRecords.map(async (dnsRecord) => {
+                const { id, username, type, name, value, user } = dnsRecord;
```
Destructure in the args and skip a step

```ts
dnsRecords.map(async ({ id, username, type, name, value, user }) => {
```

SerpentBytes replied:
Agreed.

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,105 @@
+
+import type { ExpirationStatusPayload as ExpirationRequestPayload } from '../notifications/expiration-notification.server';
+
+const { EXPIRATION_JOB_REPEAT_FREQUENCY, JOB_REMOVAL_FREQUENCY } = process.env;
```
Set defaults on these so we don't have to provide in env if we don't want to override.

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,105 @@
+
+import { addNotification } from '../notifications/notifications.server';
+import { deleteDnsRequest } from '../dns/dns-flow.server';
+import { RecordType as RecType } from '../notifications/expiration-notification.server';
```
Why change the name here?

SerpentBytes replied:
When I tried importing RecordType, I saw that the variable name was not unique; it's used in some Prisma code files and other places. To avoid any future naming conflicts, I thought to rename it.

humphd replied:
got it.

SerpentBytes replied:
For now, it's not causing a naming conflict, so I have removed the alias.

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,105 @@
+            lt: new Date(),
+          },
+        },
+        select: {
```
This probably isn't needed (i.e., you can get all the data vs trimming)

humphd reviewed (changes requested):
This is looking good.  A few things I noticed on first reading.

SerpentBytes commented:
@humphd, I have updated the code based on your initial feedback.

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,100 @@
+    jobName,
+    { type },
+    {
+      repeat: { every: Number(EXPIRATION_REPEAT_FREQUENCY) || 24 * 60 * 60 * 1000 },
```
I don't think you can use `Number()` here, since this is a cron string, right?

SerpentBytes replied:
We are not using the cron pattern anymore since not everyone is familiar with it. Therefore, we are casting the value of the environment variable into a Number, as "every" accepts a value in milliseconds.

humphd replied:
Above you have this:

```ini
# Background jobs
EXPIRATION_REPEAT_FREQUENCY=24 * 60 * 60 * 1000
JOB_REMOVAL_FREQUENCY=7 * 24 * 60  * 60 * 1000
```

Should you change that if you're not using CRON any more?  Doing `Number('7 * 24 * 60  * 60 * 1000')` will give `NaN`.

SerpentBytes replied:
Right, the quotes.

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,100 @@
+    {
+      repeat: { every: Number(EXPIRATION_REPEAT_FREQUENCY) || 24 * 60 * 60 * 1000 },
+      removeOnComplete: { age: Number(JOB_REMOVAL_FREQUENCY) || JOB_REMOVAL_INTERVAL },
+      removeOnFail: { age: Number(JOB_REMOVAL_FREQUENCY) || JOB_REMOVAL_INTERVAL },
```
Same question here

SerpentBytes replied:
see: https://github.com/DevelopingSpace/starchart/pull/357#discussion_r1140502435

SerpentBytes commented:
@humphd I have removed the `if` check and updated the notifications job-related `.env` variables values so casting to Number from String does not result in `NaN`.

humphd commented on .env.example:
```diff
@@ -30,3 +30,7 @@ SMTP_PORT=1025
 SAML_IDP_METADATA_PATH=config/idp-metadata-dev.xml
+
+# Background jobs
+EXPIRATION_REPEAT_FREQUENCY=86400000
```
What you can do here is leave a comment that explains what the humber means:

```
# 24 * 7 * ... = 86400000
EXPIRATION_REPEAT_FREQUENCY=86400000
```

Best of both worlds.

humphd replied:
Also, when you have a number like this, it's good to name it with units: `EXPIRATION_REPEAT_FREQUENCY_SECONDS` or `_MS` etc

SerpentBytes replied:
Agreed. It's helpful to know what the numbers mean.

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,98 @@
+      case RecordType.DnsRecord:
+        try {
+          logger.info('process DNS record expiration');
+          let dnsRecords = await getExpiredRecords(type);
```
So this can return a non-array?  That makes sense why you were wanting to check the `dnsRecords?.length`:

```ts
const dnsRecords = await getExpiredRecords(type);
if(!dnsRecords) {
  return;
}
return Promise.all(dnsRecords.map(...))
```

SerpentBytes replied:
Yes, it can be potentially null or undefined.

humphd reviewed (changes requested):
Looks good.  A few things I notice and I think this is ready.

SerpentBytes commented:
Updated the code with helpful comments and added if condition `return` if we get `null/undefined` back from the query that fetches expired DNS records.

dadolhay commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,101 @@
+const expirationRequestQueueName = 'expiration-request';
+
+// query to fetch expired records from database based on type
+const getExpiredRecords = (type: RecordType) => {
```
I'd suggest to create two separate functions, one for records, one for certificates. I'd probably move this functionality to the records and certificates model files too.

humphd replied:
Agree, this can move up a level in our API.

dadolhay commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,101 @@
+    jobName,
+    { type },
+    {
+      repeat: { every: Number(EXPIRATION_REPEAT_FREQUENCY_MS) || 24 * 60 * 60 * 1000 },
```
I'd make these mandatory and throw an exception if they are missing

SerpentBytes replied:
Do we still need to throw an exception if we provide a hard-coded default value?

humphd replied:
We already have too many config options to maintain.  If we can provide a sane default and override when we want to change, I'd prefer that.

dadolhay commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,101 @@
+    { type },
+    {
+      repeat: { every: Number(EXPIRATION_REPEAT_FREQUENCY_MS) || 24 * 60 * 60 * 1000 },
+      removeOnComplete: { age: Number(JOB_REMOVAL_FREQUENCY_MS) || JOB_REMOVAL_INTERVAL },
```
@humphd Should we do this in other queues too? seems like a good idea to clean stuff out of redis after a while

SerpentBytes replied:
There's an issue filed already. See #342.

dadolhay commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,101 @@
+}
+
+// worker definition
+const expirationRequestWorker = new Worker<ExpirationRequestPayload>(
```
What is the rationale in combining record and cert expiration? Wouldn't it be cleaner to make them separate?
Or do them both in one queue job (eliminating type)?

Either way would simplify the code quite a bit (especially the second)

SerpentBytes replied:
Agreed. It would be cleaner to do them in a single queue job.

SerpentBytes replied:
Let me know if [063424f](https://github.com/DevelopingSpace/starchart/pull/357/commits/063424f13bffda2bd529b0e1087f62a2f9d3b110) is the change you were requesting.

dadolhay replied:
What I suggest is to do away with `type`, i.e., this `case RecordType.DnsRecord:` would go away, and this worker would first process the DNS records, then would delete expired certificates from the DB, it would not take the "type" attribute at all

SerpentBytes replied:
Thanks for clarifying.

humphd commented on .env.example:
```diff
@@ -30,3 +30,9 @@ SMTP_PORT=1025
 SAML_IDP_METADATA_PATH=config/idp-metadata-dev.xml
+
+# Background jobs
+# 24 * 60 * 60 * 1000 = 604800000 (24 hours)
```
Awesome!

humphd commented on app/queues/common/expiration-request.server.ts:
```diff
@@ -0,0 +1,101 @@
+import type { ExpirationStatusPayload as ExpirationRequestPayload } from '../notifications/expiration-notification.server';
+
+const { EXPIRATION_REPEAT_FREQUENCY_MS, JOB_REMOVAL_FREQUENCY_MS } = process.env;
+const JOB_REMOVAL_INTERVAL = 7 * 24 * 60 * 60 * 1000;
```
Add a `_MS` to this as well, and a comment indicating what this magic number equals in hours/days or whatever.

dadolhay reviewed (dismissed):
Regardless, I'm going to approve this, if we decide to, we can make that change later, or you can incorporate it int this pr as well. Whatever makes more sense to you.

SerpentBytes commented:
@dadolhay I misunderstood you earlier. I have updated the code to process DNS records without using `type`.

dadolhay reviewed (approved):
Looks awesome!
